### PR TITLE
feat: add opkl-updater cli app

### PR DESCRIPTION
**Pull request by mrjoelkamp**
## Summary
* Adds `opkl-updater` cli app to manage the OIDC Public Key Ledger (OPKL)

Most of the important logic is in the following files:
* [cmd/update.go](https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/cmd/update.go)
* [pkg/ledger/ledger.go](https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/pkg/ledger/ledger.go)
* [pkg/ledger/issuer.go](https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/pkg/ledger/issuer.go)
* [pkg/ledger/types.go](https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/pkg/ledger/types.go)
* [pkg/ledger/utils.go](https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/pkg/ledger/utils.go)

## Testing
In Work

## Whats Next
* Git client to make branch and push changes to TUF repo
* GHA to run Dockerfile as workflow

## Example
```bash
make build
./bin/opkl-updater update https://token.actions.githubusercontent.com
```
output:
* creates or updates `targets/opkl/issuers.json` provider index for new identity providers https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/targets/opkl/issuers.json#L1-L19
* creates or updates `targets/opkl/token.actions.githubusercontent.com/pkl.json` ledger file index for all active public keys https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/targets/opkl/token.actions.githubusercontent.com/pkl.json#L1-L16
* generates ledger files referenced by the index
https://github.com/mrjoelkamp/tuf-pk-ledger/blob/cee78488a202b513d25ae0bf4b05fc2be48981cc/targets/opkl/token.actions.githubusercontent.com/47b91b6b0b44d189d9cd5ce8cea7ada0f8422f32e1cdc07874bdfe7699b513e1.json#L1-L16
* detects key rotation and updates the ledger file with the `exp` timestamp and the ledger file index status
[test: key rotation before](https://github.com/mrjoelkamp/tuf-pk-ledger/commit/da381ada28a669950f1957dfc9b7219eb4e63394)
[test: key rotation after](https://github.com/mrjoelkamp/tuf-pk-ledger/commit/fe9c1616cf2b56ebd39c7da7148f5cee22c65f86)

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
+package log
```
Comment (jonnystoten) [inline log/log.go:1]: Can probably just use https://pkg.go.dev/log/slog for logging, it's pretty similar

File: pkg/ledger/ledger.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+				indexItem.Status = StatusArchived
+				pklIdx.Items[id] = indexItem
+			}
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func reconcileExpectedJWK(jwk JWK, activeJWKs map[string]PklIndexItem) {
```
Comment (jonnystoten) [inline pkg/ledger/ledger.go]: I would inline this function to make it more obvious that the map is being mutated

File: pkg/ledger/ledger.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		err = writeJSONFile(opIdxItem.Path, pklIdx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func detectRotatedJWK(pklIdx PklIndex, expectedActiveJWKs map[string]PklIndexItem, timestamp int64) (bool, error) {
```
Comment (jonnystoten) [inline pkg/ledger/ledger.go]: Maybe `handleRotatedJWK`? `expectedActiveJWKs` already contains the rotated ones at this point right?

**Review by jonnystoten — COMMENTED** [review]
Generally looking good! Let's sync later to go over some of the Go stuff in detail.

We should move this into the `docker` namespace too.

File: log/log.go (hunk 1 above)
Comment (mrjoelkamp) [reply log/log.go:1]: Nice! I wasn't up to speed on slog essentially replacing logrus. Most of the boilerplate logging,  cobra, and viper was setup using cookiecutter (https://github.com/lacion/cookiecutter-golang). I'll refactor this to use slog, thanks!

File: pkg/ledger/ledger.go (hunk 2 above)
Comment (mrjoelkamp) [reply pkg/ledger/ledger.go]: [ca3ef02](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/ca3ef02e83ba631630cbf021f60a324533ed4b33)

File: pkg/ledger/ledger.go (hunk 3 above)
Comment (mrjoelkamp) [reply pkg/ledger/ledger.go]: [ca3ef02](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/ca3ef02e83ba631630cbf021f60a324533ed4b33)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+## Get Started!
+
+Ready to contribute? Here's how to set up `opkl-updater` for local development.
+
+1. Fork the `opkl-updater` repo on GitHub.
+2. Clone your fork locally::
+```bash
+    $ git clone git@github.com:your_name_here/opkl-updater.git
+```
```
Comment (whalelines) [inline CONTRIBUTING.md]: Using this syntax tends to not render properly, even on GitHub with its rather forgiving Markdown parser. Specifically, the preformated section is not indented:

---

2. Clone your fork locally:
```bash
    $ git clone git@github.com:your_name_here/opkl-updater.git
```

---

To correct, you have two options:

Option 1: Forgo syntax highlighting by indenting eight spaces

<pre>
2.  Clone you fork locally:

        $ git clone …
</pre>

which gives you

---

2.  Clone you fork locally:

        $ git clone …

---

Option 2: Indent the triple backticks four spaces

<pre>
2.  Clone you fork locally:

    ```bash
    $ git clone …
    ```
</pre>

which gives you

---

2.  Clone you fork locally:

    ```bash
    $ git clone …
    ```

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Build Stage
+FROM golang:1.21 AS build-stage
```
Comment (whalelines) [inline Dockerfile]: Should this be `golang:1.21-alpine`?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+COPY --from=build-stage /go/src/opkl-updater/bin/opkl-updater \
+/go/src/opkl-updater/config.opkl-updater.yaml /opt/opkl-updater/bin/
+RUN chmod +x /opt/opkl-updater/bin/opkl-updater
+
+# Create appuser
+RUN adduser -D -g '1000' opkl-updater
+USER opkl-updater
+
+CMD ["/opt/opkl-updater/bin/opkl-updater"]
```
Comment (whalelines) [inline Dockerfile]: End file with newline.

**Review by whalelines — COMMENTED** [review]
Not done with review, but it's a start.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+LABEL app="build-opkl-updater"
+LABEL REPO="https://github.com/mrjoelkamp/opkl-updater"
+
+ENV PROJPATH=/go/src/opkl-updater
+
+# Because of https://github.com/docker/docker/issues/14914
+ENV PATH=$PATH:$GOROOT/bin:$GOPATH/bin
+
+ADD . /go/src/opkl-updater
+WORKDIR /go/src/opkl-updater
```
Comment (whalelines) [inline Dockerfile:13]:
```suggestion
WORKDIR /go/src/opkl-updater
ADD . ./
```

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+LABEL GIT_COMMIT=$GIT_COMMIT
+LABEL VERSION=$VERSION
+
+# Because of https://github.com/docker/docker/issues/14914
+ENV PATH=$PATH:/opt/opkl-updater/bin
+
+WORKDIR /opt/opkl-updater/bin
+
+COPY --from=build-stage /go/src/opkl-updater/bin/opkl-updater \
+/go/src/opkl-updater/config.opkl-updater.yaml /opt/opkl-updater/bin/
```
Comment (whalelines) [inline Dockerfile:32]: Why would you put a YAML file in bin?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+RUN chmod +x /opt/opkl-updater/bin/opkl-updater
```
Comment (whalelines) [inline Dockerfile:33]: Unnecessary

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+# Create appuser
+RUN adduser -D -g '1000' opkl-updater
```
Comment (whalelines) [inline Dockerfile:36]: Move this before the COPY for better caching.

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	@echo '    make tag             Tag image created by package with latest, git commit and version'
+	@echo '    make test            Run tests on a compiled project.'
+	@echo '    make push            Push tagged images to registry'
+	@echo '    make clean           Clean the directory tree.'
+	@echo
+
+build:
+	@echo "building ${BIN_NAME} ${VERSION}"
+	@echo "GOPATH=${GOPATH}"
+	go build -ldflags "-X github.com/mrjoelkamp/opkl-updater/version.GitCommit=${GIT_COMMIT}${GIT_DIRTY} -X github.com/mrjoelkamp/opkl-updater/version.BuildDate=${BUILD_DATE}" -o bin/${BIN_NAME}
```
Comment (whalelines) [inline Makefile:33]: Why include the build date? Makes it less reproducible.

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+get-deps:
+	dep ensure
+
+build-alpine:
+	@echo "building ${BIN_NAME} ${VERSION}"
+	@echo "GOPATH=${GOPATH}"
+	go build -ldflags '-w -linkmode external -extldflags "-static" -X github.com/mrjoelkamp/opkl-updater/version.GitCommit=${GIT_COMMIT}${GIT_DIRTY} -X github.com/mrjoelkamp/opkl-updater/version.BuildDate=${BUILD_DATE}' -o bin/${BIN_NAME}
```
Comment (whalelines) [inline Makefile:41]: Why is -static in double quotes?

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+push: tag
+	@echo "Pushing docker image to registry: latest ${VERSION} $(GIT_COMMIT)"
+	docker push $(IMAGE_NAME):$(GIT_COMMIT)
+	docker push $(IMAGE_NAME):${VERSION}
+	docker push $(IMAGE_NAME):latest
+
+clean:
+	@test ! -e bin/${BIN_NAME} || rm bin/${BIN_NAME}
+
+test:
```
Comment (whalelines) [inline Makefile:62]: Should test depend on build?

File: cmd/update.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+		}
+	},
+}
+
+func UpdateLedger(cmd *cobra.Command, args []string) error {
+	err := ledger.Update(args[0])
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (whalelines) [inline cmd/update.go:29]:
```suggestion
	return ledger.Update(args[0])
```

File: config.opkl-updater.yaml [hunk 15]
```diff
@@ -0,0 +1 @@
+loglevel: info
```
Comment (whalelines) [inline config.opkl-updater.yaml:1]: Add newline

File: CONTRIBUTING.md (hunk 4 above)
Comment (mrjoelkamp) [reply CONTRIBUTING.md]: [9e1dfd0](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/9e1dfd0ffdf1d46f76efb1a243b9d94470f361a7)

File: Dockerfile (hunk 6 above)
Comment (mrjoelkamp) [reply Dockerfile]: [9e1dfd0](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/9e1dfd0ffdf1d46f76efb1a243b9d94470f361a7)

File: Dockerfile (hunk 5 above)
Comment (mrjoelkamp) [reply Dockerfile]: [9e1dfd0](https://github.com/mrjoelkamp/tuf-pk-ledger/pull/5/commits/9e1dfd0ffdf1d46f76efb1a243b9d94470f361a7)

**Comment by mrjoelkamp** [comment]
Moved to https://github.com/docker/jwkstuf
(currently not public, but will be soon)

**Comment by mrjoelkamp** [comment]
> Not done with review, but it's a start.

@whalelines if you pick it back up see https://github.com/docker/jwkstuf/pull/1

**Review by whalelines — COMMENTED** [review]
This already got merged, but I'll submit what I had before noticing that.

File: Dockerfile (hunk 7 above)
Comment (mrjoelkamp) [reply Dockerfile:13]: A lot of this Dockerfile was generated with a [cookiecutter template](https://github.com/lacion/cookiecutter-golang). I haven't spent much time on cleaning it up yet, but thank you for the suggestions!

File: Dockerfile (hunk 8 above)
Comment (mrjoelkamp) [reply Dockerfile:32]: I had a hard dependency on this config file and needed a quick way to test it. It is no longer required so I can drop it completely.

File: Makefile (hunk 13 above)
Comment (mrjoelkamp) [reply Makefile:62]: not necessary, `go test` compiles the source

```
Go test runs in two different modes:

The first, called local directory mode, occurs when go test is
invoked with no package arguments (for example, 'go test' or 'go
test -v'). In this mode, go test compiles the package sources and
tests found in the current directory and then runs the resulting
test binary. In this mode, caching (discussed below) is disabled.
After the package test finishes, go test prints a summary line
showing the test status ('ok' or 'FAIL'), package name, and elapsed
time.

The second, called package list mode, occurs when go test is invoked
with explicit package arguments (for example 'go test math', 'go
test ./...', and even 'go test .'). In this mode, go test compiles
and tests each of the packages listed on the command line. If a
package test passes, go test prints only the final 'ok' summary
line. If a package test fails, go test prints the full test output.
If invoked with the -bench or -v flag, go test prints the full
output even for passing package tests, in order to display the
requested benchmark results or verbose logging. After the package
tests for all of the listed packages finish, and their output is
printed, go test prints a final 'FAIL' status if any package test
has failed.
```

File: Makefile (hunk 11 above)
Comment (mrjoelkamp) [reply Makefile:33]: removed, it was part of the `version` command but that boilerplate isn't needed and it isn't very helpful for the reason you pointed out

**Comment by mrjoelkamp** [comment]
> This already got merged, but I'll submit what I had before noticing that.

@whalelines All good, most of them are resolved in https://github.com/docker/jwkstuf/pull/1/commits/0cdc042242e89e857a3e9019649a2aea98ea8178

File: Makefile (hunk 12 above)
Comment (mrjoelkamp) [reply Makefile:41]: its a value for `-extldflags`, doesn't seem like it is necessary but maybe it is for more than one `extldflag`. Part of the template. Tested it without the `""s` and it works fine, so I can drop them

Comment (mrjoelkamp) [reply Makefile:41]: the static linking is required though

Comment (mrjoelkamp) [reply Makefile:41]: https://github.com/docker/jwkstuf/pull/1/commits/002645889b71048e128b112fdbf170c6b724805a

Comment (whalelines) [reply Makefile:41]: The static linking was required when you switched from a golang based on Debian to a runtime based on alpine. I would bet it isn't necessary now that it is all alpine.

**Pull request merged**